### PR TITLE
CF Metrics, Tests, and Fixes

### LINKED ISSUE #29: (cf-sbd-server) publish to cloudflare and run sanity tests against production
DEPENDS ON https://github.com/holochain/sbd/issues/28

Now that we have a locally tested cf sbd server, we need to test running it on the actual cloudflare system

- publish it
- execute some sanity tests to ensure it works as desired (we can't run sbd-o-bahn against it, since that requires being able to start/stop servers willy-nilly)
- manually run `hc_demo_cli` against production cloudflare SBD server
- documentation for how to go about testing this, including how to set up a dev instance to iterate quickly against, and how to verify that a merge to production `main` deployed successfully and can be verified.

### PR DESCRIPTION
**Pull request by neonphog**
RESOLVES #29 

- Switches to the newly-recommended vitest-pool-workers wrangler test methodology
- Adds basic bytes received (and connection count) metrics, and api-key based metrics reporting
- Total rewrite of rate limit code, and tests
- Dead-status early return check in signal Durable Object

### CODE REVIEW

File: ts/sbd-server/src/common.ts [hunk 1]
```diff
@@ -0,0 +1,108 @@
+import { fromUint8Array, toUint8Array } from 'js-base64';
```
Comment (neonphog) [inline ts/sbd-server/src/common.ts:1]: The contents are this file are just consolidated from other files, no code changes here.

File: ts/sbd-server/src/signal.ts [hunk 2]
```diff
@@ -0,0 +1,407 @@
...
+   * The worker http fetch endpoint above forwards the request here.
+   * This function performs some checks, then upgrades the connection
+   * to a websocket.
+   */
+  async fetch(request: Request): Promise<Response> {
+    if (this.status !== 200) {
+      return new Response(JSON.stringify({ err: 'dead' }), {
+        status: this.status,
+      });
+    }
```
Comment (neonphog) [inline ts/sbd-server/src/signal.ts:86]: added these checks that short-circuit cpu time when the object has already errored / closed

File: ts/sbd-server/wrangler.toml [hunk 3]
```diff
@@ -1,10 +1,19 @@
...
 route = "sbd.holo.host/*"
 account_id = "18ff2b4e6205b938652998cfca0d8cff"
+minify = true
 workers_dev = false
 
+#[limits]
+#cpu_ms = 5000
+
+kv_namespaces = [
+  { binding = "SBD_COORDINATION", id = "cde857dd4ea745fea55f80daa52a6d5c" }
```
Comment (neonphog) [inline ts/sbd-server/wrangler.toml:14]: This kv store is where metrics are stored.

File: ts/sbd-server/src/index.ts [hunk 4]
```diff
@@ -91,20 +29,103 @@ export default {
...
 
       if (method !== 'GET') {
-        throw err('expected GET', 400);
+        throw common.err('expected GET', 400);
       }
 
-      const { pubKeyStr } = parsePubKey(url.pathname);
+      let pathParts = url.pathname.split('/');
+
+      if (pathParts.length === 4 && pathParts[1] === 'metrics') {
```
Comment (neonphog) [inline ts/sbd-server/src/index.ts:37]: here's the new metrics entrypoint

File: ts/sbd-server/README.md [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+`$(url)/metris/$(whoami)/$(uuidgen)`
+
+E.g. `https://sbd.holo.host/metrics/NEONPHOG/d6d38e16-9fe2-4cbf-b66d-e49775064d59`
+
+Should give you something like:
+
+```
+# HELP client.count active client count
+# TYPE client.count guage
+client.count 4
```
Comment (ThetaSinner) [inline ts/sbd-server/README.md:28]: Are these recorded somewhere? Otherwise it's just a point in time and it'd be useful to have this recorded over time

**Review by ThetaSinner — APPROVED** [review]
Looking good! and pointed a wind tunnel test at it this morning, seems to be working nicely :)

File: ts/sbd-server/README.md (hunk 5 above)
Comment (neonphog) [reply ts/sbd-server/README.md:28]: The intention would be for these to be scraped by a grafana somewhere. It'd be too expensive to aggregate them in cloudflare itself.

File: ts/sbd-server/src/prom.ts [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+  }
+  out.push('}');
+  return out.join('');
+}
+
+/**
+ * Hand-rolled simplistic prometheus metrics renderer.
+ * All the npm modules I could find out there depended on nodejs apis.
+ * I just need a simple protocol renderer that can be used in cloudflare.
+ */
```
Comment (jost-s) [inline ts/sbd-server/src/prom.ts:28]:
```suggestion
/**
 * Hand-rolled simplistic prometheus metrics renderer.
 */
 
/*
 * All the npm modules I could find out there depended on nodejs apis.
 * I just need a simple protocol renderer that can be used in cloudflare.
 */
```

**Review by jost-s — APPROVED** [review]

File: ts/sbd-server/src/prom.ts (hunk 6 above)
Comment (neonphog) [reply ts/sbd-server/src/prom.ts:28]:
```suggestion
// All the npm modules I could find out there depended on nodejs apis.
// I just need a simple protocol renderer that can be used in cloudflare.

/**
 * Hand-rolled simplistic prometheus metrics renderer.
 */
```

**Pull request merged**